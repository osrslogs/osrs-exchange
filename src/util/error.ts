/* eslint-disable max-classes-per-file */

/**
 * A Error with data details. `data` will containing the data
 * that made the error occur.
 */
class DataError extends Error {
  data: string;

  constructor(msg: string, data: string) {
    super(msg);
    this.data = data;
  }
}

/**
 * Error used when a html page has unexpected structure, e.g: Could not parse element.
 * The raw html can be found in the `data` property.
 */
export class InvalidHtmlError extends DataError {}

/**
 * A generic HTTP error, default status 500.
 * Used to map HTTP status codes to errors.
 */
export class HttpError extends Error {
  status: number;

  constructor(msg: string, status = 500) {
    super(msg);
    this.status = status;
  }
}

export class NotFoundError extends HttpError {
  constructor(msg: string) {
    super(msg, 403);
  }
}

export class ServiceUnavailableError extends HttpError {
  constructor(msg: string) {
    super(msg, 503);
  }
}
