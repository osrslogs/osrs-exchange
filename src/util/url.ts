const BASE_URL = 'http://services.runescape.com/m=itemdb_oldschool';

export const buildItemsPageUrl = (alpha: string, page: number): string => {
  return `${BASE_URL}/api/catalogue/items.json?category=1&alpha=${alpha}&page=${page}`;
};

export const buildItemPriceGraphUrl = (id: number): string => {
  return `${BASE_URL}/api/graph/${id}.json`;
};

/**
 * Builds the item url for the given item id. Note: It will redirect 302->200
 * due to not providing the name in the URL.
 *
 * @param {number} id The id of the item to lookup
 *
 * @returns {string} The url
 */
export const buildItemUrl = (id: number): string => {
  return `${BASE_URL}/viewitem?obj=${id}`;
};
