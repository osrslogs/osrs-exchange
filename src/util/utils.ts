import jsonItems from './data/items.json';
import { JsonItem, ItemName } from '../types';

const items: JsonItem = jsonItems as JsonItem;

/*
const getKeyByValue = (obj: object, value: unknown): unknown => {
  return Object.keys(obj).find(key => obj[key] === value);
};
*/

/**
 * Lookups the item id of an item by name
 *
 * @param {ItemName} name The item name
 *
 * @returns {number | undefined} The item id or undefined if not found
 */
export const getItemId = (name: ItemName): number | undefined => {
  return items[name.toLowerCase()];
};

/**
 * Converts a string, with assumed numbers, to number by removing
 * spaces and commas.
 *
 * @param {string} number
 *
 * @returns {number}
 */
export const cleanNumber = (number: string): number => {
  return Number(number.replace(/[ ,]+/g, ''));
};
