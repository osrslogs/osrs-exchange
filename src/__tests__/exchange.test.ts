import axios from 'axios';
import { mocked } from 'ts-jest/utils';
import { ItemName, Item, ItemPriceGraph, ItemDetails } from '../types';
import Exchange from '..';
import { NotFoundError } from '../util/error';
import htmlConfig from '../util/test-util/html-mock.json';
import graphConfig from '../util/test-util/graph-mock.json';
import itemPageConfig from '../util/test-util/api-item-page-mock.json';

jest.mock('axios');
const mockAxios = mocked(axios, true);
const itemId = 4151;
const itemName: ItemName = 'abyssal whip';

describe('constructor', () => {
  const defaultUserAgent = 'osrs-exchange';

  it('creates an object with UA config', () => {
    const expected = {
      userAgent: 'hello world',
    };

    const result: Exchange = new Exchange(expected);

    expect(result.config).toEqual(expected);
  });

  it('creates an object with timeout config', () => {
    const expected = {
      timeout: 123,
    };

    const result: Exchange = new Exchange(expected);

    expect(result.config.timeout).toEqual(expected.timeout);
    expect(result.config.userAgent).toEqual(defaultUserAgent);
  });

  it('creates an object with proxy config', () => {
    const expected = {
      proxy: {
        host: '123.123.123.123',
        port: 1234,
      },
    };

    const result: Exchange = new Exchange(expected);

    expect(result.config.proxy).toEqual(expected.proxy);
    expect(result.config.userAgent).toEqual(defaultUserAgent);
  });

  it('creates an object without config', () => {
    const expected = {
      userAgent: defaultUserAgent,
    };

    const result: Exchange = new Exchange();

    expect(result.config).toEqual(expected);
  });
});

describe('getItemById', () => {
  const exchange: Exchange = new Exchange();

  it('throws NotFoundError if id is not found', async () => {
    mockAxios.get.mockRejectedValueOnce({
      response: {
        status: 404,
      },
    });

    const result: Promise<Item> = exchange.getItemById(itemId);

    await expect(result).rejects.toThrow(NotFoundError);
  });

  it('returns item when the id is valid', async () => {
    mockAxios.get.mockResolvedValueOnce({
      data: htmlConfig.validPage,
    });
    const item: Item = await exchange.getItemById(itemId);
    expect(item).toMatchSnapshot();
  });
});

describe('getItemByName', () => {
  const exchange: Exchange = new Exchange();

  it('throws NotFoundError if name is not found', async () => {
    mockAxios.get.mockRejectedValueOnce({
      response: {
        status: 404,
      },
    });

    const result: Promise<Item> = exchange.getItemByName(itemName);

    await expect(result).rejects.toThrow(NotFoundError);
  });

  it('returns item when the name is valid', async () => {
    mockAxios.get.mockResolvedValueOnce({
      data: htmlConfig.validPage,
    });
    const item: Item = await exchange.getItemByName(itemName);
    expect(item).toMatchSnapshot();
  });

  it('throws NotFoundError if the name does not match an id', async () => {
    const result: Promise<Item> = exchange.getItemByName('invalid name' as ItemName);

    await expect(result).rejects.toThrow(NotFoundError);
  });
});

describe('getItemPriceGraphById', () => {
  const exchange: Exchange = new Exchange();

  it('throws NotFoundError if id is not found', async () => {
    mockAxios.get.mockRejectedValueOnce({
      response: {
        status: 404,
      },
    });

    const result: Promise<ItemPriceGraph> = exchange.getItemPriceGraphById(itemId);

    await expect(result).rejects.toThrow(NotFoundError);
  });

  it('returns item when the id is valid', async () => {
    mockAxios.get.mockResolvedValueOnce({
      data: graphConfig.valid,
    });
    const result: ItemPriceGraph = await exchange.getItemPriceGraphById(itemId);
    expect(result).toMatchSnapshot();
  });
});

describe('getItemPriceGraphByName', () => {
  const exchange: Exchange = new Exchange();

  it('throws NotFoundError if id is not found', async () => {
    mockAxios.get.mockRejectedValueOnce({
      response: {
        status: 404,
      },
    });

    const result: Promise<ItemPriceGraph> = exchange.getItemPriceGraphByName(itemName);

    await expect(result).rejects.toThrow(NotFoundError);
  });

  it('returns item when the id is valid', async () => {
    mockAxios.get.mockResolvedValueOnce({
      data: graphConfig.validTwo, // why does it have to be diff value? Does request cache previous?
    });
    const result: ItemPriceGraph = await exchange.getItemPriceGraphByName(itemName);
    expect(result).toMatchSnapshot();
  });

  it('throws NotFoundError if the name does not match an id', async () => {
    const result: Promise<ItemPriceGraph> = exchange.getItemPriceGraphByName('invalid name' as ItemName);

    await expect(result).rejects.toThrow(NotFoundError);
  });
});

describe('getItemPage', () => {
  const exchange: Exchange = new Exchange();

  it('returns item details array when valid alpha is given', async () => {
    mockAxios.get.mockResolvedValueOnce({
      data: itemPageConfig.valid,
    });

    const result: ItemDetails[] = await exchange.getItemPage('a');

    expect(result).toMatchSnapshot();
  });

  it('returns item details array when valid alpha and page is given', async () => {
    mockAxios.get.mockResolvedValueOnce({
      data: itemPageConfig.valid,
    });

    const result: ItemDetails[] = await exchange.getItemPage('b', 6);

    expect(result).toMatchSnapshot();
  });
});
