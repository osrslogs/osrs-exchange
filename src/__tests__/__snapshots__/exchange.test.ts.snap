// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getItemById returns item when the id is valid 1`] = `
Object {
  "description": "A weapon from the abyss.",
  "icon": "http://services.runescape.com/m=itemdb_oldschool/1583157682291_obj_big.gif?id=4151",
  "id": 4151,
  "members": true,
  "name": "Abyssal whip",
  "price": Object {
    "change": Object {
      "day180": -5981,
      "day30": -5981,
      "day90": -5981,
      "today": -5981,
    },
    "current": 2388108,
  },
}
`;

exports[`getItemByName returns item when the name is valid 1`] = `
Object {
  "description": "A weapon from the abyss.",
  "icon": "http://services.runescape.com/m=itemdb_oldschool/1583157682291_obj_big.gif?id=4151",
  "id": 4151,
  "members": true,
  "name": "Abyssal whip",
  "price": Object {
    "change": Object {
      "day180": -5981,
      "day30": -5981,
      "day90": -5981,
      "today": -5981,
    },
    "current": 2388108,
  },
}
`;

exports[`getItemPage returns item details array when valid alpha and page is given 1`] = `
Array [
  Object {
    "description": "world",
    "icon": "some-url",
    "id": 1234,
    "members": false,
    "name": "hello",
  },
  Object {
    "description": "world",
    "icon": "some-url",
    "id": 1234,
    "members": true,
    "name": "hello",
  },
]
`;

exports[`getItemPage returns item details array when valid alpha is given 1`] = `
Array [
  Object {
    "description": "world",
    "icon": "some-url",
    "id": 1234,
    "members": false,
    "name": "hello",
  },
  Object {
    "description": "world",
    "icon": "some-url",
    "id": 1234,
    "members": true,
    "name": "hello",
  },
]
`;

exports[`getItemPriceGraphById returns item when the id is valid 1`] = `
Object {
  "average": Object {
    "2019-09-05T00:00:00.000Z": 2528557,
    "2019-09-06T00:00:00.000Z": 2527956,
  },
  "daily": Object {
    "2019-09-05T00:00:00.000Z": 2490649,
    "2019-09-06T00:00:00.000Z": 2492999,
  },
}
`;

exports[`getItemPriceGraphByName returns item when the id is valid 1`] = `
Object {
  "average": Object {
    "2019-09-05T00:00:00.000Z": 1,
    "2019-09-06T00:00:00.000Z": 1,
  },
  "daily": Object {
    "2019-09-05T00:00:00.000Z": 1,
    "2019-09-06T00:00:00.000Z": 1,
  },
}
`;
